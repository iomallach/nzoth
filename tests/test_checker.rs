mod helpers;
use helpers::{SnapshotErrors, create_source_file};
use nzoth::{checker::Checker, parser::Parser};

fn test_valid_cases(glob_pattern: &str) {
    insta::glob!(glob_pattern, |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = create_source_file(path, contents.as_str());

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(parser.errors().len(), 0);

        let mut checker = Checker::new(&source);
        checker.typecheck_program(program);
        let checked_program = checker.comp_unit;

        let output = format!("{:#?}", checked_program);
        insta::with_settings!({
            omit_expression => true,
            input_file => path,
            snapshot_suffix => path.file_name().unwrap().to_str().unwrap(),
        }, {
            insta::assert_snapshot!(output);
        });
    })
}

fn test_invalid_cases(glob_pattern: &str) {
    insta::glob!(glob_pattern, |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = create_source_file(path, contents.as_str());

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let mut checker = Checker::new(&source);
        checker.typecheck_program(program);

        insta::with_settings!({
            omit_expression => true,
            input_file => path,
            snapshot_suffix => path.file_name().unwrap().to_str().unwrap(),
        }, {
            insta::assert_snapshot!(SnapshotErrors(&checker.errors));
        });
    })
}

#[test]
fn test_valid_let_declarations() {
    test_valid_cases("resources/checker_cases/valid_let_declarations.nz");
}

#[test]
fn test_invalid_let_declarations() {
    test_invalid_cases("resources/checker_cases/invalid_let_declarations.nz");
}

#[test]
fn test_invalid_expressions() {
    test_invalid_cases("resources/checker_cases/invalid_expressions.nz");
}
