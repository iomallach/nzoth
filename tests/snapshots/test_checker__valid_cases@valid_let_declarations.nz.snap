---
source: tests/test_checker.rs
input_file: tests/resources/checker_cases/valid_let_declarations.nz
---
CompilationUnit {
    scope: Scope {
        parent: None,
        functions: {
            "foo": CheckedFuncDeclaration {
                name: "foo",
                parameters: [],
                body: CheckedFunctionBody {
                    statements: [
                        CheckedLetVarDeclaration(
                            CheckedLetVarDeclaration {
                                name: "bar",
                                ty: 0,
                                span: Span {
                                    start: Position {
                                        offset: 138,
                                        line: 9,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 149,
                                        line: 9,
                                        column: 13,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                            NumericLiteral(
                                Integer(
                                    2,
                                ),
                                0,
                                Span {
                                    start: Position {
                                        offset: 148,
                                        line: 9,
                                        column: 12,
                                    },
                                    end: Position {
                                        offset: 149,
                                        line: 9,
                                        column: 13,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ),
                        Return(
                            CheckedReturn {
                                expression: Unit(
                                    Span {
                                        start: Position {
                                            offset: 0,
                                            line: 0,
                                            column: 0,
                                        },
                                        end: Position {
                                            offset: 0,
                                            line: 0,
                                            column: 0,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                                span: Span {
                                    start: Position {
                                        offset: 0,
                                        line: 0,
                                        column: 0,
                                    },
                                    end: Position {
                                        offset: 0,
                                        line: 0,
                                        column: 0,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                        ),
                    ],
                    span: Span {
                        start: Position {
                            offset: 134,
                            line: 8,
                            column: 13,
                        },
                        end: Position {
                            offset: 152,
                            line: 10,
                            column: 1,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                },
                return_type: 3,
                span: Span {
                    start: Position {
                        offset: 121,
                        line: 8,
                        column: 0,
                    },
                    end: Position {
                        offset: 124,
                        line: 8,
                        column: 3,
                    },
                    file: FileId(
                        0,
                    ),
                },
            },
            "main": CheckedFuncDeclaration {
                name: "main",
                parameters: [],
                body: CheckedFunctionBody {
                    statements: [
                        CheckedLetVarDeclaration(
                            CheckedLetVarDeclaration {
                                name: "foo",
                                ty: 0,
                                span: Span {
                                    start: Position {
                                        offset: 25,
                                        line: 1,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 36,
                                        line: 1,
                                        column: 13,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                            NumericLiteral(
                                Integer(
                                    1,
                                ),
                                0,
                                Span {
                                    start: Position {
                                        offset: 35,
                                        line: 1,
                                        column: 12,
                                    },
                                    end: Position {
                                        offset: 36,
                                        line: 1,
                                        column: 13,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ),
                        CheckedLetVarDeclaration(
                            CheckedLetVarDeclaration {
                                name: "bar",
                                ty: 0,
                                span: Span {
                                    start: Position {
                                        offset: 40,
                                        line: 2,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 58,
                                        line: 2,
                                        column: 20,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                            NumericLiteral(
                                Integer(
                                    1,
                                ),
                                0,
                                Span {
                                    start: Position {
                                        offset: 57,
                                        line: 2,
                                        column: 19,
                                    },
                                    end: Position {
                                        offset: 58,
                                        line: 2,
                                        column: 20,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ),
                        CheckedLetVarDeclaration(
                            CheckedLetVarDeclaration {
                                name: "baz",
                                ty: 2,
                                span: Span {
                                    start: Position {
                                        offset: 62,
                                        line: 3,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 75,
                                        line: 3,
                                        column: 15,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                            NumericLiteral(
                                Float(
                                    3.4,
                                ),
                                2,
                                Span {
                                    start: Position {
                                        offset: 72,
                                        line: 3,
                                        column: 12,
                                    },
                                    end: Position {
                                        offset: 75,
                                        line: 3,
                                        column: 15,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ),
                        CheckedLetVarDeclaration(
                            CheckedLetVarDeclaration {
                                name: "foobar",
                                ty: 2,
                                span: Span {
                                    start: Position {
                                        offset: 79,
                                        line: 4,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 104,
                                        line: 4,
                                        column: 27,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                            NumericLiteral(
                                Float(
                                    7.3,
                                ),
                                2,
                                Span {
                                    start: Position {
                                        offset: 101,
                                        line: 4,
                                        column: 24,
                                    },
                                    end: Position {
                                        offset: 104,
                                        line: 4,
                                        column: 27,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ),
                        Return(
                            CheckedReturn {
                                expression: NumericLiteral(
                                    Integer(
                                        1,
                                    ),
                                    0,
                                    Span {
                                        start: Position {
                                            offset: 115,
                                            line: 5,
                                            column: 9,
                                        },
                                        end: Position {
                                            offset: 116,
                                            line: 5,
                                            column: 10,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                                span: Span {
                                    start: Position {
                                        offset: 108,
                                        line: 5,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 116,
                                        line: 5,
                                        column: 10,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            },
                        ),
                    ],
                    span: Span {
                        start: Position {
                            offset: 21,
                            line: 0,
                            column: 21,
                        },
                        end: Position {
                            offset: 119,
                            line: 6,
                            column: 1,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                },
                return_type: 0,
                span: Span {
                    start: Position {
                        offset: 0,
                        line: 0,
                        column: 0,
                    },
                    end: Position {
                        offset: 3,
                        line: 0,
                        column: 3,
                    },
                    file: FileId(
                        0,
                    ),
                },
            },
        },
        var_decls: {},
    },
    functions: [
        CheckedFuncDeclaration {
            name: "main",
            parameters: [],
            body: CheckedFunctionBody {
                statements: [
                    CheckedLetVarDeclaration(
                        CheckedLetVarDeclaration {
                            name: "foo",
                            ty: 0,
                            span: Span {
                                start: Position {
                                    offset: 25,
                                    line: 1,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 36,
                                    line: 1,
                                    column: 13,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        NumericLiteral(
                            Integer(
                                1,
                            ),
                            0,
                            Span {
                                start: Position {
                                    offset: 35,
                                    line: 1,
                                    column: 12,
                                },
                                end: Position {
                                    offset: 36,
                                    line: 1,
                                    column: 13,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                    ),
                    CheckedLetVarDeclaration(
                        CheckedLetVarDeclaration {
                            name: "bar",
                            ty: 0,
                            span: Span {
                                start: Position {
                                    offset: 40,
                                    line: 2,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 58,
                                    line: 2,
                                    column: 20,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        NumericLiteral(
                            Integer(
                                1,
                            ),
                            0,
                            Span {
                                start: Position {
                                    offset: 57,
                                    line: 2,
                                    column: 19,
                                },
                                end: Position {
                                    offset: 58,
                                    line: 2,
                                    column: 20,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                    ),
                    CheckedLetVarDeclaration(
                        CheckedLetVarDeclaration {
                            name: "baz",
                            ty: 2,
                            span: Span {
                                start: Position {
                                    offset: 62,
                                    line: 3,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 75,
                                    line: 3,
                                    column: 15,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        NumericLiteral(
                            Float(
                                3.4,
                            ),
                            2,
                            Span {
                                start: Position {
                                    offset: 72,
                                    line: 3,
                                    column: 12,
                                },
                                end: Position {
                                    offset: 75,
                                    line: 3,
                                    column: 15,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                    ),
                    CheckedLetVarDeclaration(
                        CheckedLetVarDeclaration {
                            name: "foobar",
                            ty: 2,
                            span: Span {
                                start: Position {
                                    offset: 79,
                                    line: 4,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 104,
                                    line: 4,
                                    column: 27,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        NumericLiteral(
                            Float(
                                7.3,
                            ),
                            2,
                            Span {
                                start: Position {
                                    offset: 101,
                                    line: 4,
                                    column: 24,
                                },
                                end: Position {
                                    offset: 104,
                                    line: 4,
                                    column: 27,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                    ),
                    Return(
                        CheckedReturn {
                            expression: NumericLiteral(
                                Integer(
                                    1,
                                ),
                                0,
                                Span {
                                    start: Position {
                                        offset: 115,
                                        line: 5,
                                        column: 9,
                                    },
                                    end: Position {
                                        offset: 116,
                                        line: 5,
                                        column: 10,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                            span: Span {
                                start: Position {
                                    offset: 108,
                                    line: 5,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 116,
                                    line: 5,
                                    column: 10,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                    ),
                ],
                span: Span {
                    start: Position {
                        offset: 21,
                        line: 0,
                        column: 21,
                    },
                    end: Position {
                        offset: 119,
                        line: 6,
                        column: 1,
                    },
                    file: FileId(
                        0,
                    ),
                },
            },
            return_type: 0,
            span: Span {
                start: Position {
                    offset: 0,
                    line: 0,
                    column: 0,
                },
                end: Position {
                    offset: 3,
                    line: 0,
                    column: 3,
                },
                file: FileId(
                    0,
                ),
            },
        },
        CheckedFuncDeclaration {
            name: "foo",
            parameters: [],
            body: CheckedFunctionBody {
                statements: [
                    CheckedLetVarDeclaration(
                        CheckedLetVarDeclaration {
                            name: "bar",
                            ty: 0,
                            span: Span {
                                start: Position {
                                    offset: 138,
                                    line: 9,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 149,
                                    line: 9,
                                    column: 13,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        NumericLiteral(
                            Integer(
                                2,
                            ),
                            0,
                            Span {
                                start: Position {
                                    offset: 148,
                                    line: 9,
                                    column: 12,
                                },
                                end: Position {
                                    offset: 149,
                                    line: 9,
                                    column: 13,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                    ),
                    Return(
                        CheckedReturn {
                            expression: Unit(
                                Span {
                                    start: Position {
                                        offset: 0,
                                        line: 0,
                                        column: 0,
                                    },
                                    end: Position {
                                        offset: 0,
                                        line: 0,
                                        column: 0,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                            span: Span {
                                start: Position {
                                    offset: 0,
                                    line: 0,
                                    column: 0,
                                },
                                end: Position {
                                    offset: 0,
                                    line: 0,
                                    column: 0,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                    ),
                ],
                span: Span {
                    start: Position {
                        offset: 134,
                        line: 8,
                        column: 13,
                    },
                    end: Position {
                        offset: 152,
                        line: 10,
                        column: 1,
                    },
                    file: FileId(
                        0,
                    ),
                },
            },
            return_type: 3,
            span: Span {
                start: Position {
                    offset: 121,
                    line: 8,
                    column: 0,
                },
                end: Position {
                    offset: 124,
                    line: 8,
                    column: 3,
                },
                file: FileId(
                    0,
                ),
            },
        },
    ],
    types: [
        BuiltIn(
            Int,
        ),
        BuiltIn(
            Bool,
        ),
        BuiltIn(
            Float,
        ),
        BuiltIn(
            Unit,
        ),
    ],
}
